### PR TITLE
Dashboard: Change tracking simplified 

### PR DESCRIPTION
**Pull request by torkelo**
A poc of a simple change tracker based on just incrementing a user change counter from places that perform user-initiated changes. We really need to explore how to make DashboardModel and PanelModel immutable (except for PanelQueryRunner & event emitters), think we are close to be able to explore that (some challenges with Angular plugin support but I think we can bridge that). Given that change is coming I opted to keep this change as stupidly simple as possible. 

Pros
* No "Unsaved changes" modal triggered due to panel migrations

Cons
* Might miss incrementing this for a new option 

Thoughts? 

Todo
* [x] Handle panel updates (already done in previous PRs) 
* [x] Handle annotation query updates
* [x] Handle General setting changes
* [x] Handle variable updates #41143
* [ ] Handle dashboard link updates 

Anything missing from todo?

### CODE REVIEW

**Review by hugohaggmark — COMMENTED** [review]
I like the approach. It's a combination of user activated dashboard changes and panel changes, right?

Regarding the dashboard settings, maybe we should try to move those to Redux as an extension of this PR?

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 1]
```diff
@@ -106,6 +106,7 @@ export class DashboardModel {
   iteration?: number;
   declare meta: DashboardMeta;
   events: EventBusExtended;
+  unsavedUserChanges: number;
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard/state/DashboardModel.ts:109]: yea, good idea

**Comment by torkelo** [comment]
@hugohaggmark 
> I like the approach. It's a combination of user activated dashboard changes and panel changes, right?

Yea exactly. 

> Regarding the dashboard settings, maybe we should try to move those to Redux as an extension of this PR?

Want to explore dashboard immutable state changes in a separate PR, unsure what approach to take there and what the end goal will look like.

**Comment by torkelo** [comment]
@hugohaggmark looking at where to call incrementUnsavedUserChange for variable changes, we either pass it to the EditorList and Editor and call it from all the event callbacks (ie where the redux actions are called from). Or we do it inside the redux actions. But those redux actions have no access to dashboard / root state. So another option is to add a "hasUserChange" on VariableModel (that we do not persist), that get's set to true in actions like changeVariableProp, changeVariableType, changeVariableOrder, duplicateVariable and removeVariable. Thoughts?

**Comment by hugohaggmark** [comment]
> @hugohaggmark looking at where to call incrementUnsavedUserChange for variable changes, we either pass it to the EditorList and Editor and call it from all the event callbacks (ie where the redux actions are called from). Or we do it inside the redux actions. But those redux actions have no access to dashboard / root state. So another option is to add a "hasUserChange" on VariableModel (that we do not persist), that get's set to true in actions like changeVariableProp, changeVariableType, changeVariableOrder, duplicateVariable and removeVariable. Thoughts?

I like the latter approach. Adding that prop to variable model should be easy enough and setting the prop in the actions, too. Also, all existing reducer tests should make it easier 🤞

**Review by dprokop — COMMENTED** [review]
Looks good, I like how simplicity of it all.

**Review by ashharrison90 — APPROVED** [review]
lgtm 👍  it does feel like it could be easy to forget to include this in a callback, but i can't think of a suggestion to remedy that 😂

**Comment by hugohaggmark** [comment]
@torkelo I opened https://github.com/grafana/grafana/pull/41143 for the variables part

**Pull request closed without merging**